Outline.md

```diff
@@ -0,0 +1,137 @@
+
+
+
+## Prerequisites 
+
+To get 100% benefited from this course you must have some previous skills on programming. You need -
+
+-   Working knowledge of JavaScript and ES6+ (ESNext)
+-   Asynchronous Programming in JavaScript
+-   JavaScript Promise and Async Await
+-   Basic Understanding of Web
+
+
+
+## Lessons
+
+I will update describe every chapter later after your confirmation. 
+
+
+
+1.  Introduction
+    1.  Course Introduction (Slide)
+    2.  Course Curriculum Explanation (Slide + Screen)
+    3.  How to Use This Course (Slide)
+2.  Understand the Concept of API
+    1.  What is API (Slide)
+    2.  What is REST API (Slide)
+    3.  Why do We Need API (Slide)
+    4.  Six Constraints of REST API (Slide)
+    5.  HTTP, HTTP Methods and HTTP Status Codes (Slide)
+    6.  REST API vs SDK (Slide)
+    7.  Advantage and Disadvantage of REST API (Slide)
+    8.  REST API Alternatives (Slide)
+3.  Setting A Fresh Express Application
+    1.  Install NodeJS (screen)
+    2.  Create NodeJS Application (screen)
+    3.  Create an Express Server (screen)
+    4.  Install lowdb (json database) (screen)
+    5.  Install Necessary Middleware
+4.  Install and Explore Postman
+    1.  Download and Install Postman (screen)
+    2.  Overview of Postman (screen)
+5.  Create Our First REST API
+    1.  Explain The Todo Application (screen)
+    2.  Create A POST Route (screen)
+    3.  Create GET Routes (screen)
+    4.  Create PUT & PATCH Routes (screen)
+    5.  Create DELETE Route (screen)
+6.  CRUD Operations for Recipe
+    1.  Setup Database (Screen)
+    2.  Implement Create Recipe (Screen)
+    3.  Implement Get Recipes (Screen)
+    4.  Implement Get Recipe By Id (Screen)
+    5.  Implement Update Recipe (Screen)
+    6.  Implement Delete Recipe (Screen)
+7.  CRUD Operation for Review
+    1.  Understand The Relation Between Recipe and Review (Screen)
+    2.  Implement Create Review (Screen)
+    3.  Implement Get Review (Screen)
+    4.  Implement Get Review By Id (Screen)
+    5.  Implement Update Review (Screen)
+    6.  Implement Delete Review (Screen)
+8.  Understand JWT & Authentication
+    1.  How to Protect API (Screen)
+    2.  What is JWT (Screen)
+    3.  Structure of JWT (Screen)
+    4.  How Does JWT Authentication Work (Screen)
+    5.  Create User Model (Screen)
+    6.  Create Registration Routes (Screen)
+    7.  Create Login Routes (Screen)
+9.  Authentication using PassportJS
+    1.  Create Authentication Middleware (Screen)
+    2.  Understand PassportJS (Screen)
+    3.  Implement PassportJS Local Strategy (Screen)
+10.  REST API Best Practices
+     1.  Open API Specifications (Screen)
+     2.  API Best Practices (Slide)
+11.  API Versioning and Documentation
+     1.  How to Manage Version of API (Screen)
+     2.  Introduction to Swagger (Screen)
+     3.  Implement Swagger in Our Application (Screen)
+12.  API Testing using JEST & Supertest
```

Comment:
We should include some content here on the test pyramid: https://martinfowler.com/articles/practical-test-pyramid.html#:~:text=The%20%22Test%20Pyramid%22%20is%20a,put%20it%20into%20practice%20properly.

Particularly relevant here is:
1. API System Test - Call the actual API in its actual environment, no mocking whatsoever. In my experience this is typically done with Jest + Node Fetch calls.
2. API Service Test - Call the API, but mocking all external dependencies (e.g. other APIs being called). In my experience this is done with Jest + Supertest (as you outline here)
3. Unit Tests - Calls functions and methods directly, testing internal functionality (provides safety when refactoring). In my experience this is typically done in Jest alone.